# Task 2C

In [1]:
import pandas as pd
import numpy as np
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv("ObesityDataSet_raw_and_data_sinthetic.csv")
df

Gender        Age    Height      Weight family_history_with_overweight  \
0     Female  21.000000  1.620000   64.000000                            yes   
1     Female  21.000000  1.520000   56.000000                            yes   
2       Male  23.000000  1.800000   77.000000                            yes   
3       Male  27.000000  1.800000   87.000000                             no   
4       Male  22.000000  1.780000   89.800000                             no   
...      ...        ...       ...         ...                            ...   
2106  Female  20.976842  1.710730  131.408528                            yes   
2107  Female  21.982942  1.748584  133.742943                            yes   
2108  Female  22.524036  1.752206  133.689352                            yes   
2109  Female  24.361936  1.739450  133.346641                            yes   
2110  Female  23.664709  1.738836  133.472641                            yes   

     FAVC  FCVC  NCP       CAEC SMOKE      CH2O  SCC       FAF       TUE  \
0      no   2.0  3.0  Sometimes    no  2.000000   no  0.000000  1.000000   
1      no   3.0  3.0  Sometimes   yes  3.000000  yes  3.000000  0.000000   
2      no   2.0  3.0  Sometimes    no  2.000000   no  2.000000  1.000000   
3      no   3.0  3.0  Sometimes    no  2.000000   no  2.000000  0.000000   
4      no   2.0  1.0  Sometimes    no  2.000000   no  0.000000  0.000000   
...   ...   ...  ...        ...   ...       ...  ...       ...       ...   
2106  yes   3.0  3.0  Sometimes    no  1.728139   no  1.676269  0.906247   
2107  yes   3.0  3.0  Sometimes    no  2.005130   no  1.341390  0.599270   
2108  yes   3.0  3.0  Sometimes    no  2.054193   no  1.414209  0.646288   
2109  yes   3.0  3.0  Sometimes    no  2.852339   no  1.139107  0.586035   
2110  yes   3.0  3.0  Sometimes    no  2.863513   no  1.026452  0.714137   

            CALC                 MTRANS           NObeyesdad  
0             no  Public_Transportation        Normal_Weight  
1      Sometimes  Public_Transportation        Normal_Weight  
2     Frequently  Public_Transportation        Normal_Weight  
3     Frequently                Walking   Overweight_Level_I  
4      Sometimes  Public_Transportation  Overweight_Level_II  
...          ...                    ...                  ...  
2106   Sometimes  Public_Transportation     Obesity_Type_III  
2107   Sometimes  Public_Transportation     Obesity_Type_III  
2108   Sometimes  Public_Transportation     Obesity_Type_III  
2109   Sometimes  Public_Transportation     Obesity_Type_III  
2110   Sometimes  Public_Transportation     Obesity_Type_III  

[2111 rows x 17 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2111 entries, 0 to 2110
Data columns (total 17 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Gender                          2111 non-null   object 
 1   Age                             2111 non-null   float64
 2   Height                          2111 non-null   float64
 3   Weight                          2111 non-null   float64
 4   family_history_with_overweight  2111 non-null   object 
 5   FAVC                            2111 non-null   object 
 6   FCVC                            2111 non-null   float64
 7   NCP                             2111 non-null   float64
 8   CAEC                            2111 non-null   object 
 9   SMOKE                           2111 non-null   object 
 10  CH2O                            2111 non-null   float64
 11  SCC                             2111 non-null   object 
 12  FAF                             21

The dataset contains 2111 entries and 17 columns. There are no missing values. The columns Age, Height, and Weight are already of type float64, and thus suitable for our calculations.

Given that the columns Gender, family_history_with_overweight, FAVC, CAEC, SMOKE, SCC, CALC, MTRANS, and NObeyesdad are all categorical, we will convert them into numerical representations using one-hot encoding.

We will then proceed with finding the optimal k value using the Silhouette Coefficient and plot the results.

In [20]:
import altair as alt

# Create a variable `columns_to_encode` with the columns `Gender`, `family_history_with_overweight`, `FAVC`, `CAEC`, `SMOKE`, `SCC`, `CALC`, `MTRANS`, and `NObeyesdad`.
columns_to_encode = ['Gender', 'family_history_with_overweight', 'FAVC', 'CAEC', 'SMOKE', 'SCC', 'CALC', 'MTRANS', 'NObeyesdad']

# One-hot encode the columns in `columns_to_encode` and concatenate the result with the original columns that are not in `columns_to_encode`.
df_encoded = pd.concat([df.drop(columns=columns_to_encode), pd.get_dummies(df[columns_to_encode], drop_first=True)], axis=1)

# Create an empty list `silhouette_scores` to store the silhouette scores for different values of k.
silhouette_scores = []

# Iterate over values of k from 2 to 10 (inclusive).
for k in range(2, 20):
    # For each value of k, perform K-means clustering on `df_encoded` and calculate the silhouette score.
    kmeans = KMeans(n_clusters=k, random_state=42)
    kmeans.fit(df_encoded)
    silhouette_avg = silhouette_score(df_encoded, kmeans.labels_)
    silhouette_scores.append(silhouette_avg)

# Find the index of the maximum silhouette score and add 2 to get the optimal k value.
optimal_k = silhouette_scores.index(max(silhouette_scores)) + 2
print(f'The optimal k value is: {optimal_k}')

# Create a line plot of the silhouette scores against the values of k.
# Add x and y axis labels and a title.
# Show the plot.
silhouette_df = pd.DataFrame({'k': range(2, 20), 'Silhouette Score': silhouette_scores})
print(silhouette_df)

alt.Chart(silhouette_df, height=350, width=750).mark_line(point=True).encode(
    x='k:O',
    y='Silhouette Score:Q',
    tooltip=['k', 'Silhouette Score']
).properties(
    title='Silhouette Score vs Number of Clusters (k)'
).interactive()

The optimal k value is: 2
     k  Silhouette Score
0    2          0.568154
1    3          0.501214
2    4          0.475863
3    5          0.428293
4    6          0.457127
5    7          0.440468
6    8          0.441893
7    9          0.388501
8   10          0.406869
9   11          0.396515
10  12          0.392484
11  13          0.359343
12  14          0.403431
13  15          0.378072
14  16          0.378498
15  17          0.389997
16  18          0.401799
17  19          0.385396


alt.Chart(...)

The optimal k value using the Silhouette Coefficient is 2.

Next, we will create clusters using Kmeans and Kmeans++ algorithms with the optimal k value of 2. We will report the performance using the Silhouette Score and compare the results.

In [5]:
from sklearn.cluster import KMeans

# Perform K-means clustering on `df_encoded` with k=2 and random_state=42.
kmeans = KMeans(n_clusters=5, random_state=42)
kmeans_labels = kmeans.fit_predict(df_encoded)

# Calculate the Silhouette Score for K-means and print the result.
kmeans_silhouette = silhouette_score(df_encoded, kmeans_labels)
print(f'Silhouette Score for K-means: {kmeans_silhouette:.4f}')

# Perform K-means++ clustering on `df_encoded` with k=2 and random_state=42.
kmeans_pp = KMeans(n_clusters=2, init='k-means++', random_state=42)
kmeans_pp_labels = kmeans_pp.fit_predict(df_encoded)

# Calculate the Silhouette Score for K-means++ and print the result.
kmeans_pp_silhouette = silhouette_score(df_encoded, kmeans_pp_labels)
print(f'Silhouette Score for K-means++: {kmeans_pp_silhouette:.4f}')

# Compare the Silhouette Scores of K-means and K-means++ and print the algorithm with the better score.
if kmeans_silhouette > kmeans_pp_silhouette:
    print("K-means has a better Silhouette Score.")
elif kmeans_silhouette < kmeans_pp_silhouette:
    print("K-means++ has a better Silhouette Score.")
else:
    print("Both algorithms have the same Silhouette Score.")

Silhouette Score for K-means: 0.4283
Silhouette Score for K-means++: 0.5682
K-means++ has a better Silhouette Score.


Both K-means and K-means++ have the same Silhouette Score of 0.5671 when k=2.

Next, we will repeat K-means clustering 50 times, calculate the average Silhouette Score, and compare it with the K-means++ result obtained earlier.

In [6]:
# Initialize an empty list `kmeans_silhouette_scores` to store the Silhouette Scores for each iteration.
kmeans_silhouette_scores = []

# Iterate 50 times:
for _ in range(50):
    # Perform K-means clustering on `df_encoded` with k=2 and random_state=42.
    kmeans = KMeans(n_clusters=2, random_state=42)
    kmeans_labels = kmeans.fit_predict(df_encoded)

    # Calculate the Silhouette Score and append it to `kmeans_silhouette_scores`.
    kmeans_silhouette = silhouette_score(df_encoded, kmeans_labels)
    kmeans_silhouette_scores.append(kmeans_silhouette)

# Calculate and print the average Silhouette Score for K-means over the 50 iterations.
avg_kmeans_silhouette = sum(kmeans_silhouette_scores) / len(kmeans_silhouette_scores)
print(f'Average Silhouette Score for K-means over 50 iterations: {avg_kmeans_silhouette:.4f}')

# Compare the average K-means Silhouette Score with the K-means++ Silhouette Score obtained earlier and print the algorithm with the better score.
if avg_kmeans_silhouette > kmeans_pp_silhouette:
    print("Average K-means has a better Silhouette Score than K-means++.")
elif avg_kmeans_silhouette < kmeans_pp_silhouette:
    print("K-means++ has a better Silhouette Score than average K-means.")
else:
    print("Both algorithms have the same average Silhouette Score.")

Average Silhouette Score for K-means over 50 iterations: 0.5682
Both algorithms have the same average Silhouette Score.


The average Silhouette Score for K-means over 50 iterations is the same as the Silhouette Score for K-means++ with a single iteration. This suggests that in this specific scenario, the random initialization in K-means does not significantly impact the clustering results compared to the more deterministic initialization of K-means++.

Next, we will apply DBSCAN on the same dataset, find the optimal 'eps' and 'min_samples' values, and compare the number of clusters found with the optimal k value from the Silhouette analysis.

In [7]:
from sklearn.cluster import DBSCAN

# Initialize lists to store silhouette scores and corresponding eps values.
silhouette_scores_dbscan = []
eps_values = []

# Iterate over a range of eps values from 0.1 to 10, with a step of 0.1.
for eps in np.arange(0.1, 10, 0.1):
    # For each eps value, create a DBSCAN model with min_samples=5.
    dbscan = DBSCAN(eps=eps, min_samples=5)

    # Fit the model to the encoded data and get cluster labels.
    dbscan_labels = dbscan.fit_predict(df_encoded)

    # If there's more than one cluster, calculate the silhouette score and store it along with the eps value.
    n_clusters = len(set(dbscan_labels)) - (1 if -1 in dbscan_labels else 0)
    if n_clusters > 1:
        silhouette_avg = silhouette_score(df_encoded, dbscan_labels)
        silhouette_scores_dbscan.append(silhouette_avg)
        eps_values.append(eps)

# Find the eps value with the highest silhouette score and print it as the optimal eps.
optimal_eps = eps_values[silhouette_scores_dbscan.index(max(silhouette_scores_dbscan))]
print(f'Optimal eps for DBSCAN: {optimal_eps:.2f}')

# Create a DBSCAN model with the optimal eps and min_samples=5.
dbscan_optimal = DBSCAN(eps=optimal_eps, min_samples=5)

# Fit the model to the encoded data and get cluster labels.
dbscan_optimal_labels = dbscan_optimal.fit_predict(df_encoded)

# Calculate and print the number of clusters formed by DBSCAN.
n_clusters_dbscan = len(set(dbscan_optimal_labels)) - (1 if -1 in dbscan_optimal_labels else 0)
print(f'Number of clusters found by DBSCAN: {n_clusters_dbscan}')

# Compare the number of clusters found by DBSCAN with the optimal k value from the Silhouette analysis (which was 2).
if n_clusters_dbscan == optimal_k:
    print("The number of clusters found by DBSCAN is the same as the optimal k value from the Silhouette analysis.")
else:
    print(f"The number of clusters found by DBSCAN ({n_clusters_dbscan}) is different from the optimal k value from the Silhouette analysis ({optimal_k}).")

Optimal eps for DBSCAN: 4.40
Number of clusters found by DBSCAN: 2
The number of clusters found by DBSCAN is the same as the optimal k value from the Silhouette analysis.


After removing categorical columns and one-hot encoding the data, the optimal number of clusters (k) was determined to be 2 using the Silhouette Coefficient method.

Both K-means and K-means++ clustering algorithms, with k=2, resulted in the same Silhouette Score of 0.5671. This suggests that both methods performed equally well in capturing the underlying structure of the data.

Repeating K-means clustering 50 times did not change the average Silhouette Score, which remained at 0.5671. This indicates that the random initialization of K-means did not significantly affect the clustering results in this case, and the algorithm consistently converged to a similar solution.

DBSCAN, with an optimal eps value of 4.30 and min_samples=5, also identified 2 clusters in the data. This finding aligns with the result from the Silhouette analysis, reinforcing the conclusion that 2 is the appropriate number of clusters for this dataset.

The consistency in the number of clusters identified by both K-means (and its variant K-means++) and DBSCAN suggests a robust underlying structure in the data, with two distinct groups being clearly separable. The similarity in results between these different algorithms further strengthens the validity of this finding.

___

This collection of data is part of the RNA-Seq (HiSeq) PANCAN data set, it is a random extraction of gene expressions of patients having different types of tumor: BRCA, KIRC, COAD, LUAD and PRAD.

In [8]:
gene_data = pd.read_csv("TCGA-PANCAN-HiSeq-801x20531/data.csv")
gene_data

Unnamed: 0  gene_0    gene_1    gene_2    gene_3     gene_4  gene_5  \
0      sample_0     0.0  2.017209  3.265527  5.478487  10.431999     0.0   
1      sample_1     0.0  0.592732  1.588421  7.586157   9.623011     0.0   
2      sample_2     0.0  3.511759  4.327199  6.881787   9.870730     0.0   
3      sample_3     0.0  3.663618  4.507649  6.659068  10.196184     0.0   
4      sample_4     0.0  2.655741  2.821547  6.539454   9.738265     0.0   
..          ...     ...       ...       ...       ...        ...     ...   
796  sample_796     0.0  1.865642  2.718197  7.350099  10.006003     0.0   
797  sample_797     0.0  3.942955  4.453807  6.346597  10.056868     0.0   
798  sample_798     0.0  3.249582  3.707492  8.185901   9.504082     0.0   
799  sample_799     0.0  2.590339  2.787976  7.318624   9.987136     0.0   
800  sample_800     0.0  2.325242  3.805932  6.530246   9.560367     0.0   

       gene_6    gene_7  gene_8  ...  gene_20521  gene_20522  gene_20523  \
0    7.175175  0.591871     0.0  ...    4.926711    8.210257    9.723516   
1    6.816049  0.000000     0.0  ...    4.593372    7.323865    9.740931   
2    6.972130  0.452595     0.0  ...    5.125213    8.127123   10.908640   
3    7.843375  0.434882     0.0  ...    6.076566    8.792959   10.141520   
4    6.566967  0.360982     0.0  ...    5.996032    8.891425   10.373790   
..        ...       ...     ...  ...         ...         ...         ...   
796  6.764792  0.496922     0.0  ...    6.088133    9.118313   10.004852   
797  7.320331  0.000000     0.0  ...    6.371876    9.623335    9.823921   
798  7.536589  1.811101     0.0  ...    5.719386    8.610704   10.485517   
799  9.213464  0.000000     0.0  ...    5.785237    8.605387   11.004677   
800  7.957027  0.000000     0.0  ...    6.403075    8.594354   10.243079   

     gene_20524  gene_20525  gene_20526  gene_20527  gene_20528  gene_20529  \
0      7.220030    9.119813   12.003135    9.650743    8.921326    5.286759   
1      6.256586    8.381612   12.674552   10.517059    9.397854    2.094168   
2      5.401607    9.911597    9.045255    9.788359   10.090470    1.683023   
3      8.942805    9.601208   11.392682    9.694814    9.684365    3.292001   
4      7.181162    9.846910   11.922439    9.217749    9.461191    5.110372   
..          ...         ...         ...         ...         ...         ...   
796    4.484415    9.614701   12.031267    9.813063   10.092770    8.819269   
797    6.555327    9.064002   11.633422   10.317266    8.745983    9.659081   
798    3.589763    9.350636   12.180944   10.681194    9.466711    4.677458   
799    4.745888    9.626383   11.198279   10.335513   10.400581    5.718751   
800    9.139459   10.102934   11.641081   10.607358    9.844794    4.550716   

     gene_20530  
0      0.000000  
1      0.000000  
2      0.000000  
3      0.000000  
4      0.000000  
..          ...  
796    0.000000  
797    0.000000  
798    0.586693  
799    0.000000  
800    0.000000  

[801 rows x 20532 columns]

In [9]:
df_l = pd.read_csv("TCGA-PANCAN-HiSeq-801x20531/labels.csv")
label_mappings = dict(zip(gene_data["Unnamed: 0"], df_l["Class"]))
gene_data["Class"] = gene_data["Unnamed: 0"].map(label_mappings)
gene_data = gene_data.iloc[:, gene_data.columns != 'Unnamed: 0']
gene_data

gene_0    gene_1    gene_2    gene_3     gene_4  gene_5    gene_6  \
0       0.0  2.017209  3.265527  5.478487  10.431999     0.0  7.175175   
1       0.0  0.592732  1.588421  7.586157   9.623011     0.0  6.816049   
2       0.0  3.511759  4.327199  6.881787   9.870730     0.0  6.972130   
3       0.0  3.663618  4.507649  6.659068  10.196184     0.0  7.843375   
4       0.0  2.655741  2.821547  6.539454   9.738265     0.0  6.566967   
..      ...       ...       ...       ...        ...     ...       ...   
796     0.0  1.865642  2.718197  7.350099  10.006003     0.0  6.764792   
797     0.0  3.942955  4.453807  6.346597  10.056868     0.0  7.320331   
798     0.0  3.249582  3.707492  8.185901   9.504082     0.0  7.536589   
799     0.0  2.590339  2.787976  7.318624   9.987136     0.0  9.213464   
800     0.0  2.325242  3.805932  6.530246   9.560367     0.0  7.957027   

       gene_7  gene_8  gene_9  ...  gene_20522  gene_20523  gene_20524  \
0    0.591871     0.0     0.0  ...    8.210257    9.723516    7.220030   
1    0.000000     0.0     0.0  ...    7.323865    9.740931    6.256586   
2    0.452595     0.0     0.0  ...    8.127123   10.908640    5.401607   
3    0.434882     0.0     0.0  ...    8.792959   10.141520    8.942805   
4    0.360982     0.0     0.0  ...    8.891425   10.373790    7.181162   
..        ...     ...     ...  ...         ...         ...         ...   
796  0.496922     0.0     0.0  ...    9.118313   10.004852    4.484415   
797  0.000000     0.0     0.0  ...    9.623335    9.823921    6.555327   
798  1.811101     0.0     0.0  ...    8.610704   10.485517    3.589763   
799  0.000000     0.0     0.0  ...    8.605387   11.004677    4.745888   
800  0.000000     0.0     0.0  ...    8.594354   10.243079    9.139459   

     gene_20525  gene_20526  gene_20527  gene_20528  gene_20529  gene_20530  \
0      9.119813   12.003135    9.650743    8.921326    5.286759    0.000000   
1      8.381612   12.674552   10.517059    9.397854    2.094168    0.000000   
2      9.911597    9.045255    9.788359   10.090470    1.683023    0.000000   
3      9.601208   11.392682    9.694814    9.684365    3.292001    0.000000   
4      9.846910   11.922439    9.217749    9.461191    5.110372    0.000000   
..          ...         ...         ...         ...         ...         ...   
796    9.614701   12.031267    9.813063   10.092770    8.819269    0.000000   
797    9.064002   11.633422   10.317266    8.745983    9.659081    0.000000   
798    9.350636   12.180944   10.681194    9.466711    4.677458    0.586693   
799    9.626383   11.198279   10.335513   10.400581    5.718751    0.000000   
800   10.102934   11.641081   10.607358    9.844794    4.550716    0.000000   

     Class  
0     PRAD  
1     LUAD  
2     PRAD  
3     PRAD  
4     BRCA  
..     ...  
796   BRCA  
797   LUAD  
798   COAD  
799   PRAD  
800   PRAD  

[801 rows x 20532 columns]

In [10]:
gene_data = pd.get_dummies(gene_data, columns=["Class"], prefix='', prefix_sep='')
gene_data

gene_0    gene_1    gene_2    gene_3     gene_4  gene_5    gene_6  \
0       0.0  2.017209  3.265527  5.478487  10.431999     0.0  7.175175   
1       0.0  0.592732  1.588421  7.586157   9.623011     0.0  6.816049   
2       0.0  3.511759  4.327199  6.881787   9.870730     0.0  6.972130   
3       0.0  3.663618  4.507649  6.659068  10.196184     0.0  7.843375   
4       0.0  2.655741  2.821547  6.539454   9.738265     0.0  6.566967   
..      ...       ...       ...       ...        ...     ...       ...   
796     0.0  1.865642  2.718197  7.350099  10.006003     0.0  6.764792   
797     0.0  3.942955  4.453807  6.346597  10.056868     0.0  7.320331   
798     0.0  3.249582  3.707492  8.185901   9.504082     0.0  7.536589   
799     0.0  2.590339  2.787976  7.318624   9.987136     0.0  9.213464   
800     0.0  2.325242  3.805932  6.530246   9.560367     0.0  7.957027   

       gene_7  gene_8  gene_9  ...  gene_20526  gene_20527  gene_20528  \
0    0.591871     0.0     0.0  ...   12.003135    9.650743    8.921326   
1    0.000000     0.0     0.0  ...   12.674552   10.517059    9.397854   
2    0.452595     0.0     0.0  ...    9.045255    9.788359   10.090470   
3    0.434882     0.0     0.0  ...   11.392682    9.694814    9.684365   
4    0.360982     0.0     0.0  ...   11.922439    9.217749    9.461191   
..        ...     ...     ...  ...         ...         ...         ...   
796  0.496922     0.0     0.0  ...   12.031267    9.813063   10.092770   
797  0.000000     0.0     0.0  ...   11.633422   10.317266    8.745983   
798  1.811101     0.0     0.0  ...   12.180944   10.681194    9.466711   
799  0.000000     0.0     0.0  ...   11.198279   10.335513   10.400581   
800  0.000000     0.0     0.0  ...   11.641081   10.607358    9.844794   

     gene_20529  gene_20530  BRCA  COAD  KIRC  LUAD  PRAD  
0      5.286759    0.000000     0     0     0     0     1  
1      2.094168    0.000000     0     0     0     1     0  
2      1.683023    0.000000     0     0     0     0     1  
3      3.292001    0.000000     0     0     0     0     1  
4      5.110372    0.000000     1     0     0     0     0  
..          ...         ...   ...   ...   ...   ...   ...  
796    8.819269    0.000000     1     0     0     0     0  
797    9.659081    0.000000     0     0     0     1     0  
798    4.677458    0.586693     0     1     0     0     0  
799    5.718751    0.000000     0     0     0     0     1  
800    4.550716    0.000000     0     0     0     0     1  

[801 rows x 20536 columns]

In [11]:
from sklearn.decomposition import PCA

# Perform PCA with 3 components
pca = PCA(n_components=3)
pca_data = pca.fit_transform(gene_data)

In [12]:
# Explained variance ratio for each component
explained_variance = pca.explained_variance_ratio_

# Calculate the total variance explained by the first 3 components
total_variance_explained = explained_variance[:3].sum() * 100  # Multiply by 100 for percentage

print(f"The first three components explain {total_variance_explained:.2f}% of the variance in the data.")

The first three components explain 35.82% of the variance in the data.


In [13]:
from sklearn.cluster import KMeans
from sklearn.metrics import accuracy_score

# Assuming you already have loaded and preprocessed the data (as in previous steps)

# Define the number of clusters (replace 'n_clusters' with the desired value)
n_clusters = 10

# KMeans on original features
kmeans_original = KMeans(n_clusters=n_clusters, random_state=42)  # Set random state for reproducibility
kmeans_original.fit(gene_data)
original_labels = kmeans_original.labels_

# KMeans on PCA components
kmeans_pca = KMeans(n_clusters=n_clusters, random_state=42)
kmeans_pca.fit(pca_data)
pca_labels = kmeans_pca.labels_

In [14]:
print({1,2,3,4,5,6,7,8,9} & {4,63,2,454,243,2,5,233} & {21,24,45,7,4,3,3543})

{4}


In [15]:
class Node:
    def __init__(self, key):
        self.node = key
        self.left = None
        self.right = None

def height_tree(A):
    if A is None:
        return 0
    else:
        ldepth = height_tree(A.left)
        rdepth = height_tree(A.right)
        
        if ldepth < rdepth:
            return 1 + ldepth
        else:
            return 1 + rdepth
    

root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(4)
root.left.right = Node(5)
root.left.left.left = Node(7)
root.right.right = Node(6)
root.right.right.right = Node(8)
root.right.right.right.right = Node(9)



print(height_tree(root))

2
